das/s3_das.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package das
+
+import (
+	"bytes"
+	"context"
+	"encoding/base32"
+	"errors"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/offchainlabs/nitro/arbstate"
+	"github.com/offchainlabs/nitro/blsSignatures"
+
+	flag "github.com/spf13/pflag"
+)
+
+var s3DasMutex sync.Mutex
+
+type S3Config struct {
+	AccessKey string `koanf:"access-key"`
+	Bucket    string `koanf:"bucket"`
+	ObjectKey string `koanf:"object-key"`
+	Region    string `koanf:"region"`
+	SecretKey string `koanf:"secret-key"`
+}
+
+var DefaultS3Config = S3Config{
+	AccessKey: "",
+	Bucket:    "",
+	ObjectKey: "",
+	Region:    "",
+	SecretKey: "",
+}
+
+type S3DataAvailabilityService struct {
+	s3Config   S3Config
+	pubKey     *blsSignatures.PublicKey
+	privKey    blsSignatures.PrivateKey
+	uploader   *manager.Uploader
+	downloader *manager.Downloader
+	signerMask uint64
+}
+
+func S3ConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.String(prefix+".access-key", DefaultS3Config.AccessKey, "S3 access key")
+	f.String(prefix+".bucket", DefaultS3Config.Bucket, "S3 bucket")
+	f.String(prefix+".object-key", DefaultS3Config.ObjectKey, "S3 object key")
+	f.String(prefix+".region", DefaultS3Config.Region, "S3 region")
+	f.String(prefix+".secret-key", DefaultS3Config.SecretKey, "S3 secret key")
+}
+
+func readKeysFromS3(s3Config S3Config, downloader *manager.Downloader) (*blsSignatures.PublicKey, blsSignatures.PrivateKey, error) {
+	pubKeyBuf := manager.NewWriteAtBuffer([]byte{})
+	_, err := downloader.Download(context.TODO(), pubKeyBuf, &s3.GetObjectInput{
```

Comment:
Here and everywhere you're using the TODO context, you can pass in the context from the Store/Retrieve

Reply:
Done.